crates/codegen/schema/src/validation/rules/references.rs

```diff
@@ -34,17 +37,13 @@ impl<'v> ReferencesCollector<'v> {
 }
 
 impl Visitor for ReferencesCollector<'_> {
-    fn visit_expression(
-        &mut self,
-        expression: &ExpressionRef,
-        reporter: &mut Reporter,
-    ) -> VisitorResponse {
-        if let ExpressionParser::Reference(reference) = &expression.parser.value {
-            let Node { syntax, value } = reference;
-            if self.definitions.expressions.contains_key(value) {
-                reporter.report(syntax, Errors::Illegal(value.to_owned()))
+    fn visit_scanner(&mut self, scanner: &ScannerRef, reporter: &mut Reporter) -> VisitorResponse {
```

Comment:
nit: the validation code is repeated now three times. I suggest moving it to a common function.